### ISSUE TITLE
Question about the calculation of N-MPJPE

### ISSUE THREAD

**Issue by leon532** [hrhodin/UnsupervisedGeometryAwareRepresentationLearning#9]
Hi,

Thank you for sharing your code.

I have a question why do you use torch.sum(torch.mul(pred_vec,gt_vec),1,keepdim=True) to calculate dot_pose_gt, but not torch.sum(torch.mul(gt_vec,gt_vec),1,keepdim=True).
https://github.com/hrhodin/UnsupervisedGeometryAwareRepresentationLearning/blob/4e121586ee2f9016a1db69de1ca2811c6744f9ec/python/losses/poses.py#L17

Could you please give me an instruction or related literature reference?

Thank you for your time, and look forward to you feedback.

**Comment by hrhodin** [comment]
It was introduced in https://arxiv.org/pdf/1803.04775.pdf
It is a least squares fit, easy to compute:
Let p be the prediction, g the GT, and s the sought scale. We find the s that minimizes
min_x ||s p - q||^2
= min_x (s p - q) * (s p - q)
Take the derivative and set to zero to find min:
2 p * (s p -g) = 0
<=> s p * p - p * g = 0
<=> s = p * g / (p * p)
where * is the dot product

**Comment by leon532** [comment]
Thanks for your detailed reply. :)

**Comment by matteorr** [comment]
@hrhodin - Following up on the question above. By looking at:

https://github.com/hrhodin/UnsupervisedGeometryAwareRepresentationLearning/blob/4e121586ee2f9016a1db69de1ca2811c6744f9ec/python/configs/train_encodeDecode_pose.py#L50-L51

It seems that your training loss is simple mean squared error computed on the normalized outputs and targets (i.e. normalized over keypoint channel), while the N-MPJPE loss is used only to measure your performance at test time. 

Is my understanding correct, or is your training loss also using the N-MPJPECriterion? If not, have you tried that?

Thanks a lot.

**Comment by leon532** [comment]
By the way, I would like to provide a more detailed instruction about the calculation of s.

Let ![formula3](https://user-images.githubusercontent.com/2986699/68636123-a27e5e00-054e-11ea-9f99-3e60f499c793.png) be the predicted and ground-truth 3D positions of human joints, the objective function of the least-squares fit should be ![formula4](https://user-images.githubusercontent.com/2986699/68636227-ea04ea00-054e-11ea-9af1-ab5a34a9cd61.png), where \|sA-B\|_{F}^{2} is the Frobenius norm.
Then the pipeline of obtaining s should be:
![formula1](https://user-images.githubusercontent.com/2986699/68635414-8ed1f800-054c-11ea-8640-e20a352c1be4.png)

Take the derivative of $s$ and set to zero to find min:
![formula2](https://user-images.githubusercontent.com/2986699/68635325-4e727a00-054c-11ea-9a11-9094454f700b.png)

Please refer to the detailed instruction of Orthogonal Procrustes problem: https://en.wikipedia.org/wiki/Orthogonal_Procrustes_problem

I hope this can make it clearer to readers.

**Comment by hrhodin** [comment]
@matteorr  No, I haven't tried, could help a bit. I was using the scaling of both input and GT becasue it is simple.
For testing, to be compatible with other metrics, we want to keep the target scale and only scale the source. Thats why I used that other loss. 

@absurdliyang Thanks for flashing it out. Note that procrustes also rotates. The one here is a simpler problem that needs no SVD etc.
